### ISSUE TITLE
torch_scatter.scatter: how is the output arranged?  It seems the ouput elements does not align with the ordered indices.

### ISSUE THREAD

**Issue by yanconglin** [pyg-team/pytorch_geometric#6859]
### 🚀 The feature, motivation and pitch

> Question: 1.how is the output arranged, in terms of order?  

Please check the toy example below (pytorch-scatter: 2.1.0, py39_torch_1.12.0_cu116, pyg)
```

import random
import numpy as np
import torch
import os
import torch_geometric as pyg

seed = 0
random.seed(seed)
np.random.seed(seed)
torch.manual_seed(seed)
# Toy example: using a simple 1d tensor for illustration: 
# In reality, src could be multi-dimension tensors, 
# as shown in  https://pytorch-scatter.readthedocs.io/en/latest/functions/scatter.html
src = torch.randint(low=0, high=10, size=(5, 3))
print(f'src: {src.numpy()}')

# expecting the output to be:
# src[unique(index)], where unique(index) follows a certain order, and the output follows exactly the same order. 

index = torch.tensor([0, 1, 1, 2, 2])
out = pyg.utils.scatter(src, index, dim=0, reduce="max")
print(f'exp 1: index and output: \n {index.numpy()}, \n {out.numpy()}')

index = torch.tensor([0, 2, 2, 1, 1])
out = pyg.utils.scatter(src, index, dim=0, reduce="max")
print(f'exp 2: index and output: \n {index.numpy()}, \n {out.numpy()}')

index = torch.tensor([1, 0, 0, 2, 2])
out = pyg.utils.scatter(src, index, dim=0, reduce="max")
print(f'exp 3: index and output: \n {index.numpy()}, \n {out.numpy()}')

index = torch.tensor([2, 0, 0, 1, 1])
out = pyg.utils.scatter(src, index, dim=0, reduce="max")
print(f'exp 4: index and output: \n {index.numpy()}, \n {out.numpy()}')
```
```
# output: 
src: [[4 9 3]
 [0 3 9]
 [7 3 7]
 [3 1 6]
 [6 9 8]]
exp 1: index and output: 
 [0 1 1 2 2], 
 [[4 9 3] # idx=0
 [7 3 9]  # idx=1
 [6 9 8]] # idx=2
exp 2: index and output: 
 [0 2 2 1 1], 
 [[4 9 3]  # idx=0
 [6 9 8]  # idx=1
 [7 3 9]]  # idx=2
exp 3: index and output: 
 [1 0 0 2 2], 
 [[7 3 9] # idx=0
 [4 9 3] # idx= 1
 [6 9 8]] # idx=2
exp 4: index and output: 
 [2 0 0 1 1], 
 [[7 3 9]  # idx=0
 [6 9 8] # idx=1
 [4 9 3]] # idx=2

```

I see it seems th eouput is ordered in such a way: 
say firtst argsort(index), and then enumerate over sorted(index) and do the reduction operation.
Am I understanding correctly? It seems more like clarification problem instead of a feature request. 
Could you please also refer me to the cuda implementation of scatter? Appreciate it.

**Comment by rusty1s** [comment]
Can you clarify on your issues with `scatter`? `scatter_reduce` is a different operation than `index_select` (which you are referring to here) - all the outputs look absolutely correct to me: `src_i` gets placed into the `out` vector at index `index_i`.

This is also exactly what you need to tackle your use-case.

**Comment by yanconglin** [comment]
Thank u for the reply. I have updated the question.  I was looking for clarification now.

**Comment by rusty1s** [comment]
If your `index` is one-dimensional, but you are operating on a multi-dimensional tensor `src`, then the `index` will be broadcasted. Hope this clarifies your doubts.

**Comment by yanconglin** [comment]
thank u for the swift reply. unfortunately, it seems the discussion goes off track (my apologies for any confusion). can you pls refer me to the source cuda code then I will figure it out on my own.

**Comment by rusty1s** [comment]
You can take a look at
* https://pytorch-scatter.readthedocs.io/en/latest/functions/scatter.html
* https://github.com/rusty1s/pytorch_scatter/blob/master/csrc/cpu/scatter_cpu.cpp

Still happy to answer any questions, just not totally sure about your doubts yet.

**Comment by yanconglin** [comment]
appreciate it!

**Comment by rusty1s** [comment]
I think the text is correct. `out` has the same number of dimensions, and the shape matches to the shape of `src` except for in dimension `dim`, where `out.size(dim)` is defined through `index.max() + 1`.

**Comment by yanconglin** [comment]
after some tests,  confusion resolved. thank you for the reply.
